BUG: read_csv throws UnicodeDecodeError with unicode aliases

nateGeorge opened the pull request:
Rebased as PR https://github.com/pydata/pandas/pull/14060

---
- [x] closes #13549 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

-read_csv with engine=c throws error when encoding=UTF_16 (anything other than utf-16)
-improved nosetests and moved to in pandas/io/tests/common.py
-passes `pep8radius upstream/master --diff` and  `git diff upstream/master | flake8 --diff`
-put what's new entry in 0.19.0 in accordance with milestone posted on issue

gfyoung committed:
CLN: Remove the engine parameter in CSVFormatter and to_csv

closes #13419
xref #11274

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

see issue #13549
read_csv with engine=c throws error when encoding=UTF_16
or when encoding has _ or caps

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

see issue #13549
read_csv with engine=c throws error when encoding=UTF_16
or when encoding has _ or caps

sinhrks commented on pandas/tests/io/test_encoding_aliases.py:
```diff
@@ -0,0 +1,32 @@
+import pandas, os, nose
+
```
can u move tests to ``pandas/tests/io/parser/common.py``?

nateGeorge replied:
done

jreback replied:
might as well test for variants of UTF8 as well (eg with - and _)

nateGeorge replied:
Ok, I added utf-8.  How exactly do I run all the tests once I have it in pandas/io/tests/parser/common.py?  I tried `nosetests pandas/io/tests/parser/common.py` from root and it ran 0 tests.  When I run it with `nosetests pandas/tests/io/test_encoding_aliases.py` it works.

gfyoung replied:
The test suite is "short enough" that I believe you can run `pd.test(raise_warnings=())` and cover everything.  The way the test suite works is as follows:

The main test suite is `parsers.py`, which imports from `common.py` and a WHOLE TON of other modules in that same directory.  It is these imports that make it tricky to run individually.  If you look at the test classes, you see that they inherit from test classes like those in `common.py` to compose the test suites for each engine (`'c'` and `'python'`).

sinhrks commented on pandas/tests/io/test_encoding_aliases.py:
```diff
@@ -0,0 +1,32 @@
+                    path,
+                    engine='c',
+                    encoding=encoding)
+                print(encoding, 'succeeded with engine =', engine)
```
prepare the expected dataframe and compare the result using ``tm.assert_frame_equal``.

nateGeorge replied:
Alright, but why tm.assert_frame_equal if assert was working?

sinhrks commented on pandas/tests/io/test_encoding_aliases.py:
```diff
@@ -0,0 +1,32 @@
+                    encoding=encoding)
+                print(encoding, 'succeeded with engine =', engine)
+                test_output.append(1)
+            except UnicodeDecodeError:
```
This should not happen after the fix, correct? Pls remove.

nateGeorge replied:
Ok

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode

improved testing, added utf-8 to testing,
moved testing to pandas/io/tests/parser/common.py
see issue # 13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode

improved testing, added utf-8 to testing,
moved testing to pandas/io/tests/parser/common.py
see issue # 13549

nateGeorge committed:
Merge branch 'master' of github.com:pydata/pandas into fix/read_csv-utf-aliases

master got one commit ahead before I noticed

sinhrks commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,34 @@ def test_memory_map(self):
+
+        for encoding in test_encodings:
+            for engine in engines:
+                out = pd.io.parsers.read_csv(
```
use ``self.read_csv`` as test util automatically switches all engines (as below)

- https://github.com/pydata/pandas/blob/master/pandas/io/tests/parser/test_parsers.py#L72

nateGeorge replied:
Thanks, wasn't sure if that would work

sinhrks commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,34 @@ def test_memory_map(self):
+        # see gh issue 13549
+        engines = ['c', 'python', None]
+        path = 'test.csv'
+        expected = DataFrame({"A": [0, 1], "B": [2, 3]})
```
can we include multibytes columns / values?

sinhrks commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,34 @@ def test_memory_map(self):
+        engines = ['c', 'python', None]
+        path = 'test.csv'
+        expected = DataFrame({"A": [0, 1], "B": [2, 3]})
+        expected.to_csv(path, encoding='utf-8', index=False)
```
how about merging 2 tests as:

```
for byte in [8, 16]
   for fmt in ['utf-{0}',  'utf_{0}',  'UTF-{0}',  'UTF_{0}']
      encoding = fmt.format(byte)
      ...
```

nateGeorge replied:
ok

sinhrks commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,34 @@ def test_memory_map(self):
         out = self.read_csv(mmap_file, memory_map=True)
         tm.assert_frame_equal(out, expected)
+
+    def test_read_csv_utf_aliases():
```
self

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

read_csv with engine=c throws error when encoding=UTF_16
or when encoding has _ or uppercase
improved testing loops and added multibyte testing
see issue #13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

read_csv with engine=c throws error when encoding=UTF_16
or when encoding has _ or uppercase
improved testing loops and added multibyte testing
see issue #13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

removed `pd` from `pd.DataFrame`
see issue #13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

removed `pd` from `pd.DataFrame`
see issue #13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

fixed pep8 formatting issue
see issue #13549

nateGeorge committed:
BUG: `read_csv` throws UnicodeDecodeError with unicode aliases

fixed pep8 formatting issue
see issue #13549

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,23 @@ def test_memory_map(self):
+            expected.to_csv(path, encoding='utf-' + str(byte), index=False)
+            for fmt in ['utf-{0}', 'utf_{0}', 'UTF-{0}', 'UTF_{0}']:
+                encoding = fmt.format(byte)
+                for engine in ['c', 'python', None]:
```
you don't iterate thru the engines, this is what ``self.read_csv`` does automatically.

nateGeorge replied:
So should I just write it like

```python
for byte in [8, 16]:
	expected.to_csv(path, encoding='utf-' + str(byte), index=False)
	for fmt in ['utf-{0}', 'utf_{0}', 'UTF-{0}', 'UTF_{0}']:
		encoding = fmt.format(byte)
		result = self.read_csv(
			path,
			engine=engine,
			encoding=encoding)
		tm.assert_frame_equal(result, expected)
```

gfyoung replied:
Yes, that's correct (I didn't see this before because it was hidden away), see <a href="https://github.com/pydata/pandas/pull/13571#discussion_r70371310">below</a>.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -81,3 +81,4 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- Bug in ``read_csv()``, where aliases for utf-xx (e.g. UTF-xx, UTF_xx, utf_xx) raised UnicodeDecodeError (:issue:`13549`)
```
put in 0.18.2

jreback replied:
double backticks around ``UnicodeDecodeError``

jreback replied:
pd.read_csv()

nateGeorge replied:
Looks like 0.18.2 was moved to 0.19.0 this weekend

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,23 @@ def test_memory_map(self):
+
+    def test_read_csv_utf_aliases(self):
+        # see gh issue 13549
+        path = 'test.csv'
```
use the context manager

```
with tm.ensure_clean(path) as path:
```

remove ``os.remove(..)``

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -1469,3 +1469,23 @@ def test_memory_map(self):
+            for fmt in ['utf-{0}', 'utf_{0}', 'UTF-{0}', 'UTF_{0}']:
+                encoding = fmt.format(byte)
+                for engine in ['c', 'python', None]:
+                    out = self.read_csv(
```
out->result

nateGeorge replied:
Do you mean change variable 'out' to 'result'?  There are at least 8 other test functions in that file that use 'out' and 'expected' and do tm.assert_numpy_array_equal(out, expected)

jreback replied:
oh i c. we almost always use ``result`` not ``out`` but not a big deal. 

we NEVER use an ``assert_numpy_array_equal`` unles you are explicity comparing numpy arrays. Since you are comparing frames use ``assert_frame_equal``

jreback replied:
the prior function is explicity comparing for rec-arrays

nateGeorge replied:
Ok, I'll change the 'out's to 'result's in the file for consistency.  Is that ok to include in this pull?

gfyoung replied:
Surprised no one responded to this, but as that change is NOT directly related to your main fix, it would have been preferable to put in a separate PR (or at least in an individual commit SEPARATE from anything else).  So don't undo what you've done, but just for future reference.

nateGeorge replied:
Ok, thanks.

adneu committed:
BUG: Groupby.nth includes group key inconsistently #12839

closes #12839

Author: adneu <aneumann31@gmail.com>

Closes #13316 from adneu/12839 and squashes the following commits:

16f5cd3 [adneu] Name change
ac1851a [adneu] Added docstrings/comments, and new tests.
4d73cbf [adneu] Updated tests
9b75df4 [adneu] BUG: Groupby.nth includes group key inconsistently #12839

parthea committed:
In gbq, use googleapiclient instead of apiclient #13454 (#13458)

closes #13454

jreback committed:
COMPAT: 32-bit compat fixes mainly in testing

closes #13566
closes #13584

jorisvandenbossche committed:
RLS: switch master from 0.18.2 to 0.19.0 (#13586)

haleemur committed:
BUG: Datetime64Formatter not respecting ``formatter``

 - [x] closes #10690   - [x] tests added / passed   - [x] passes ``git
diff upstream/master | flake8 --diff``   - [x] whatsnew entry    the
Datetime64Formatter class did not accept a `formatter` argument, so
custom formatters passed in through `df.to_string` or `df.to_html`
were silently ignored.

Author: Haleemur Ali <haleemur@gmail.com>

This patch had conflicts when merged, resolved by
Committer: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Closes #13567 from haleemur/fix/dt64_outputformat and squashes the following commits:

8d84283 [Haleemur Ali] fix bug in Datetime64Formatter, which affected custom date formatted output for df.to_string, df.to_html methods

yui-knk committed:
BUG: Fix TimeDelta to Timedelta (#13600)

sinhrks committed:
BUG: DatetimeIndex - Period shows ununderstandable error

closes #13078

Author: sinhrks <sinhrks@gmail.com>

Closes #13581 from sinhrks/dti_period_error and squashes the following commits:

c957541 [sinhrks] BUG: DatetimeIndex - Period shows ununderstandable error

gfyoung committed:
ENH: add downcast to pd.to_numeric

Title is self-explanatory.  Closes #13352.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13425 from gfyoung/to-numeric-enhance and squashes the following commits:

4758dcc [gfyoung] ENH: add 'downcast' to pd.to_numeric

sinhrks committed:
CLN: remove radd workaround in ops.py

Remove workaround added in #353.

Author: sinhrks <sinhrks@gmail.com>

Closes #13606 from sinhrks/ops_radd_cln and squashes the following commits:

d873aad [sinhrks] CLN: remove radd workaround

sinhrks committed:
DEPR: rename Timestamp.offset to .freq

closes #12160

Author: sinhrks <sinhrks@gmail.com>

Closes #13593 from sinhrks/depr_timestamp_offset and squashes the following commits:

c7749d5 [sinhrks] DEPR: rename Timestamp.offset to .freq

sinhrks committed:
BUG: Block/DTI doesnt handle tzlocal properly

Author: sinhrks <sinhrks@gmail.com>

Closes #13583 from sinhrks/tzlocal and squashes the following commits:

93f59a3 [sinhrks] BUG: DTI doesnt handle tzlocal properly

sinhrks committed:
BUG: Series contains NaT with object dtype comparison incorrect (#13592)

closes #9005

sinhrks committed:
CLN/TST: Add tests for nan/nat mixed input (#13477)

closes #13467

jorisvandenbossche committed:
BUG: groupby apply on selected columns yielding scalar (GH13568) (#13585)

closes #13568

IamJeffG committed:
TST: Clean up tests of DataFrame.sort_{index,values} (#13496)

* TST: Clean up tests of DataFrame.sort_{index,values}

* Factor out Series sorting tests to own file.

* Delegate deprecated sort() and order() to their own tests.

Before this commit, the `Series.sort_values()` tests relied on deprecated
`Series.sort()` and `Series.order()` as the source of truth. However
they both merely called `Series.sort_values()` under the hood.

This commit consolidates the core test logic against `.sort_values()`
directly, while `.sort()` and `.order()` merely check for equivalence
with `.sort_values()`.

Also removes some no-op assertions that had rotted from the old days of
`sort()`/`order()`.

* Remove 'by' docstring from Series.sort_values

* Document defaults for optional sorting args

* Move more sort_values, sort_index tests to be together.

* Add test for Series.sort_index(sort_remaining=True)

* Improve `sort_values` tests when multiple `by`s

Duplicates values in the test DataFrame are necessary
to fully test this feature.

* PEP8 cleanup

* Annotate tests with GH issue

* Fix indentation - docstring string replacement

nateGeorge committed:
chore: matched master

nateGeorge committed:
DOC: add pd.read_csv bug #13549

nateGeorge committed:
DOC: add pd.read_csv bug #13549

nateGeorge committed:
TST: out-> result and tm.ensure_clean

nateGeorge committed:
TST: out-> result and tm.ensure_clean

nateGeorge committed:
TST: conform to PEP8

nateGeorge committed:
TST: conform to PEP8

gfyoung commented on pandas/io/tests/parser/common.py:
```diff
@@ -1467,5 +1469,23 @@ def test_memory_map(self):
+    def test_read_csv_utf_aliases(self):
+        # see gh issue 13549
+        path = 'test.csv'
+        expected = DataFrame({'A': [0, 1], 'B': [2, 3],
```
1) We like to have tests that are as compact as possible.  Do we really need to have this many rows for this test?  Can we get away with just one?

2) To make these tests as unit-like as possible, we would like to NOT use `to_csv` and follow the `StringIO(data)` paradigm.  I believe that is possible here because you can encode strings as `utf-8` or `utf-16`.

nateGeorge replied:
I suppose we could do one row as 

expected = pd.DataFrame({'mb_num': [4.8], 'multibyte': ['test']})

I used BytesIO because I don't think StringIO can support different encodings (I tried and wasn't able to get StringIO to work).

gfyoung commented on pandas/io/tests/parser/common.py:
```diff
@@ -1467,5 +1469,23 @@ def test_memory_map(self):
+                expected.to_csv(path, encoding='utf-' + str(byte), index=False)
+                for fmt in ['utf-{0}', 'utf_{0}', 'UTF-{0}', 'UTF_{0}']:
+                    encoding = fmt.format(byte)
+                    for engine in ['c', 'python', None]:
```
This is not necessary (nor is the `engine` keyword).  The test suite will cover both engines (and the default case is not needed here).  That's why it's `self.read_csv` and not `read_csv`.  You are in fact running the test suite TWICE for each engine.

nateGeorge replied:
Alright.

jorisvandenbossche committed:
DOC: asfreq clarify original NaNs are not filled (GH9963) (#13617)

nateGeorge committed:
TST: condense test_read_utf_aliases test

-use BytesIO instead of reading & writing file
-shorten expected DataFrame

nateGeorge committed:
TST: condense test_read_utf_aliases test

-use BytesIO instead of reading & writing file
-shorten expected DataFrame

nateGeorge committed:
Merge branch 'master' of github.com:pydata/pandas into fix/read_csv-utf-aliases

Merge latest doc commit from master.

sinhrks committed:
BUG: Invalid Timedelta op may raise ValueError

Author: sinhrks <sinhrks@gmail.com>

Closes #13624 from sinhrks/timedelta_comp and squashes the following commits:

856df95 [sinhrks] BUG: Invalid Timedelta op may raise ValueError

sinhrks committed:
CLN: Cleanup ops.py

Author: sinhrks <sinhrks@gmail.com>

Closes #13605 from sinhrks/ops_cln2 and squashes the following commits:

729997b [sinhrks] CLN: Cleanup ops.py

gfyoung committed:
CLN: Removed outtype in DataFrame.to_dict (#13627)

Follows up from #8486 in 0.15.0 by removing outtype in DataFrame.to_dict()

yui-knk committed:
CLN: Fix compile time warnings

This commit suppresses these warnings

warning: comparison of constant -1 with expression\
of type 'PANDAS_DATETIMEUNIT' is always true\
[-Wtautological-constant-out-of-range-compare]

Author: yui-knk <spiketeika@gmail.com>

Closes #13607 from yui-knk/fix_c_warning and squashes the following commits:

e9eee1d [yui-knk] CLN: Fix compile time warnings

nateGeorge committed:
Merge branch 'master' of github.com:pydata/pandas into fix/read_csv-utf-aliases

Merge master

gfyoung commented on pandas/io/tests/parser/common.py:
```diff
@@ -1467,5 +1470,16 @@ def test_memory_map(self):
+        # see gh issue 13549
+        expected = pd.DataFrame({'mb_num': [4.8], 'multibyte': ['test']})
+        for byte in [8, 16]:
+            expected.to_csv(path, encoding='utf-' + str(byte), index=False)
```
You don't need this anymore.

nateGeorge replied:
Ah yeah, sorry.  I'm testing the test in a separate file because I haven't figured out how to run the nosetests in common.py yet.

gfyoung commented on pandas/io/tests/parser/common.py:
```diff
@@ -3,6 +3,7 @@
 import csv
 import os
 import platform
+from io import BytesIO
```
Standard and safer is to import this from `pandas.compat`

nateGeorge committed:
CLN: remove unnecessary BytesIO import

nateGeorge committed:
CLN: remove unnecessary BytesIO import

nateGeorge committed:
CLN: remove unnecessary csv write line

nateGeorge committed:
CLN: remove unnecessary csv write line

jorisvandenbossche committed:
Pin IPython for doc build to 4.x (see #13639)

jreback committed:
CLN: reorg type inference & introspection

closes #12503

Author: Jeff Reback <jeff@reback.net>

Closes #13147 from jreback/types and squashes the following commits:

244649a [Jeff Reback] CLN: reorg type inference & introspection

gfyoung committed:
BLD: included pandas.api.* in setup.py (#13640)

nateGeorge committed:
Merge branch 'master' of github.com:pydata/pandas into fix/read_csv-utf-aliases

Merge new master PRs

jorisvandenbossche commented:
@nateGeorge Can you rebase this?

@gfyoung @jreback further comments?

gfyoung commented:
LGTM

nateGeorge committed:
docs: add note about read_csv() bug

see #13549

nateGeorge committed:
cln: trying to merge with master

nateGeorge committed:
CLN: merge with master

nateGeorge commented:
Ok, I think it's good now.  I was having trouble with `git --rebase` so I did a `git merge` instead.

jorisvandenbossche commented:
@nateGeorge you picked some unneeded changes in v0.19.0. 
Normally, rebasing like:

```
git fetch upstream
git rebase upstream/master
git push -f origin fix/read_csv-utf-aliases
```

should work fine. What didn't work?

nateGeorge commented:
Hmm, I was following [this guide](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request), which said to do:

```
git fetch upstream
git merge-base fix/read_csv-utf-aliases upstream/master
```
which gives a hash for the common base node of my branch and the master, then

```
git rebase -i ${HASH}
```

using that hash.  This gave me the error:

```
error: could not apply ${some commit}
```

Then I looked at the file it couldn't apply some commit to (I think v0.19.0.txt), and it had dozens of merge conflicts.  It seemed like I would have to go look up each commit to see which part of the merge to use.  I fixed one file, and then did `git rebase --continue`, and the same error came up again on another file, and I think on the same file again.  The main culprit was the whatsnew/v0.19.0.txt file.

So I guess next time I should just do

```
git fetch upstream
git rebase upstream/master
git push -f origin fix/read_csv-utf-aliases
```

and forget the whole hash thing.

jorisvandenbossche commented:
You will still have to clean up the commits now (many unrelated here, and the diff is also not fully correct)

nateGeorge commented:
Yes.  I think I'll get to it tomorrow, it looks like a giant mess now :boom:

nateGeorge committed:
Merge branch 'fix/read_csv-utf-aliases' of github.com:nateGeorge/pandas into fix/read_csv-utf-aliases

nateGeorge committed:
CLN: reset to master branch

nateGeorge committed:
Merge branch 'master' of github.com:pydata/pandas into fix/read_csv-utf-aliases

nateGeorge committed:
CLN: fix small diff from upstream/master

jorisvandenbossche commented:
@nateGeorge It seems it still didn't work well. I would do the following to fix it:

```
git checkout master
git checkout -b temp   # create a temporary branch to fix things
git merge --squash fix/read_csv-utf-aliases  ## squashes all commits of that branch that are ahead of master down into one
git checkout master
git branch -m temp fix/read_csv-utf-aliases   ## rename to orignal branch
```

jorisvandenbossche commented:
But possibly you will have to reset the last few commits, because now there is no diff anymore.

nateGeorge commented:
What I just did was delete everything locally and replace with the current upstream/master from a zip.  I couldn't figure out another way to do it for over an hour.  Thanks for the tip, maybe I'll use it in the future.  I'm working on replacing what I had in there now.

nateGeorge committed:
BUG: _read encoding fix
change encoding to lowercase
sub - for _

see #13549

nateGeorge committed:
DOC: add note on read_csv bug

see #13549

nateGeorge committed:
TST: add test for read_csv with unicode bug

see #13549

nateGeorge committed:
CLN: fix indents and spacings

nateGeorge commented:
I think it should be ready now.  I tested the test again (still haven't figured out an easy way to do that) and it passed.

jorisvandenbossche commented:
@nateGeorge There are still a lot of commits here. If you use the approach I outlined above (https://github.com/pydata/pandas/pull/13571#issuecomment-240993602), I think that should work. 
I can also clean it up when merging if you want.

nateGeorge commented:
Ok, I'll try what you advised and see what happens.

nateGeorge commented:
Hmm well it wasn't letting me overwrite the branch so I deleted it and re-created it with `git branch -m temp fix/read_csv-utf-aliases`.  Do I have to open a new PR now?

jorisvandenbossche commented:
Normally not, as long as the branch you want to push has eventually the same name (after renaming), you can just push and the PR will be updated. 
However, if you close the PR and afterwards push, github does not allow to reopen. So see if you can reopen and then push, otherwise open a new PR.

nateGeorge commented:
The 'reopen and comment' button is greyed out and says 'The fix/read_csv-utf-aliases branch was force-pushed or recreated.'  Start a new PR I guess?

jorisvandenbossche commented:
yep

nateGeorge commented:
Alright, reopened as #14060
